### ISSUE TITLE
Try supporting protocols that include `Mapped` with type union

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#10673]
The following code reports errors:
```py
from typing import Protocol
from uuid import UUID, uuid4

import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, declared_attr


class ModelBase(DeclarativeBase):
    pass


class CompareProtocol(Protocol):
    id: Mapped[int | UUID]  # Or `Mapped[int] | Mapped[UUID]`


class CompareMixin:
    def compare(self: CompareProtocol, other: CompareProtocol) -> bool:
        return self.id == other.id  # Simplified example


class IntIdMixin:
    @declared_attr
    def id(cls) -> Mapped[int]:
        return sa.orm.mapped_column(sa.Integer, primary_key=True)


class UuidIdMixin:
    @declared_attr
    def id(cls) -> Mapped[UUID]:
        return sa.orm.mapped_column(sa.UUID, primary_key=True, default=uuid4)


class MyModel(CompareMixin, IntIdMixin, ModelBase):
    pass


m1 = MyModel()
m2 = MyModel()
m1.compare(m2)  # Mypy reports two errors here
```

This errors with both mypy and pylance.

**Comment by jace** [comment]
Related Mypy bug: https://github.com/python/mypy/issues/16709

Working solution:

1. Use `declared_attr[type]` in the Protocol, not `Mapped[type]`.
2. Don't use `@declared_attr` with the decorator form. Instead, use it to create an attribute:
    ```python
    class MyMixin:
        @classmethod
        def __foo(cls) -> Mapped[...]: ...
        foo = declared_attr(__foo)
        del __foo
    ```

Possible resolutions in SQLAlchemy:

1. Make `declared_attr` generic on a covariant, matching the change to `Mapped` and related types in #10288 and #10289 (where `declared_attr` was missed).
2. Make `declared_attr` a sub-class of `Mapped` so it will match in protocols. Possibly also needed for `hybrid_property` since that's the other decorator, and decorators can't be typed like attributes.
3. Suggest the non-decorator form in documentation as a workaround when needing compatibility with Protocol. I'm not optimistic about the Mypy issue being resolved in the short term because this problem is present in all versions from 0.7 to 1.8 and appears to be fairly deep into how Mypy works. Pyright doesn't have this problem.

**Comment by CaselIT** [comment]
thanks for investigating this. if 1 works then I think it's the best solution, since it's likely that it was missed in #10288

**Comment by jace** [comment]
Yes, give me a couple days. This doesn't work at the moment, possibly because of the variance mismatch, or because declared_attr isn't a Mapped subclass. I'll investigate:

```python
    foo: Mapped[int] = declared_attr(__foo)
```

**Comment by jace** [comment]
I totally forgot about this, my bad. Looks like something has changed in Mypy and some of my previous workarounds are broken.

**Comment by sqla-tester** [comment]
**Luiz Felipe Neves** has proposed a fix for this issue in the **main** branch:

**Fixes: #10673: make declared_attr covariant** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6669

**Comment by sqla-tester** [comment]
**Luiz Felipe Neves** has proposed a fix for this issue in the **rel_2_0** branch:

**Fixes: #10673: make declared_attr covariant** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6671
